### PR TITLE
Feature/integrate sdc rabbit python library

### PR DESCRIPTION
**Pull request by Dave-Haynes**
sdx-collect should use sdc-rabbit library.

## Checklist
  - [x] CHANGELOG.md updated? (if required)

### CODE REVIEW

**Comment by necrophonic** [comment]
Need to fill in the pull request template :)

**Comment by Dave-Haynes** [comment]
[SDX Collect should use new sdc-rabbit-python library (5)](https://trello.com/c/mzvZYMoc/201-sdx-collect-should-use-new-sdc-rabbit-python-library-5)

File: app/main.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+        urls=app.settings.RABBIT_URLS,
+        queue=app.settings.RABBIT_QUARANTINE_QUEUE
+    )
+    message_consumer = MessageConsumer(
+        consumer, quarantine_publisher,
+        process=ResponseProcessor.process
+    )
+
+    try:
+        message_consumer._consumer.run()
```
Comment (JamesGardiner) [inline app/main.py]: It may change in future versions.

File: Makefile [hunk 2]
```diff
@@ -1,13 +1,19 @@
...
+build: clean
 	git clone -b 0.7.0 https://github.com/ONSdigital/sdx-common.git
-	pip install ./sdx-common
-	pip3 install -r requirements.txt
+	pip3 install ./sdx-common
 	rm -rf sdx-common
+	pip3 install -r requirements.txt
+
+dev:
+	cd .. && pip3 uninstall -y sdx-common && pip3 install -I ./sdx-common
```
Comment (insacuri) [inline Makefile]: Why does it need to be uninstalled and reinstalled?

Comment (warrenbailey) [inline Makefile]: To remove the old version and put the new one in place.

I think this can go now we've released sdx-common to pypi

File: app/response_processor.py [hunk 3]
```diff
@@ -46,41 +51,34 @@ def service_name(self, url=None):
             elif 'validate' in parts:
                 return 'SDX-VALIDATE'
         except AttributeError as e:
-            self.logger.error(e)
+            self.logger.error(str(e))
```
Comment (warrenbailey) [inline app/response_processor.py]: Why is this change needed?

File: app/response_processor.py [hunk 4]
```diff
@@ -46,41 +51,34 @@ def service_name(self, url=None):
...
         self.logger = self.logger.bind(tx_id=self.tx_id)
+        self.validate_survey(decrypted_json)
+        self.store_survey(decrypted_json)
 
-        try:
-            self.validate_survey(decrypted_json)
-        except BadMessageError:
-            # If the validation fails, the message is to be marked "invalid"
-            # and then stored. We don't then want to stop processing at this point.
-            decrypted_json['invalid'] = True
```
Comment (warrenbailey) [inline app/response_processor.py]: Don't we need this functionality?

File: app/response_processor.py [hunk 5]
```diff
@@ -46,41 +51,34 @@ def service_name(self, url=None):
...
-            self.logger.info("Invalid survey data, skipping receipting", tx_id=self.tx_id)
+        if decrypted_json.get("survey_id") != "feedback":
+            self.logger.info("Receipting survey", tx_id=self.tx_id)
```
Comment (warrenbailey) [inline app/response_processor.py]: And isn't this going to receipt even if its invalid?

File: app/response_processor.py [hunk 6]
```diff
@@ -98,53 +96,66 @@ def send_receipt(self, decrypted_json):
...
         self.logger.info("Receipt published")
 
     def decrypt_survey(self, encrypted_survey):
         self.logger.info("Decrypting survey")
-        response = self.remote_call(settings.SDX_DECRYPT_URL, data=encrypted_survey)
-        try:
-            self.response_ok(response)
-        except BadMessageError:
-            # Translate the Bad Message into a Decrypt Error to force quarantine
-            raise DecryptError
```
Comment (warrenbailey) [inline app/response_processor.py]: Again, don't we need this DecryptError otherwise we won't Quarantine

File: Makefile (hunk 2 above)
Comment (necrophonic) [inline Makefile]: The `dev` target should be removed altogether - should be the same process for dev and prod deployment

File: app/response_processor.py (hunk 6 above)
Comment (necrophonic) [inline app/response_processor.py]: Yup still need this - otherwise the next step attempts to read data from the decoded JSON (that'll fail naturally) and cause the message to be rejected altogether rather than quarantined.

File: app/response_processor.py (hunk 5 above)
Comment (necrophonic) [inline app/response_processor.py]: Also does this need the `tx_id` as it should already be bound to the logger?

File: Makefile (hunk 2 above)
Comment (JamesGardiner) [inline Makefile]: So can `clean`

File: Makefile [hunk 7]
```diff
@@ -1,13 +1,19 @@
...
 
 test:
 	pip3 install -r test_requirements.txt
 	flake8 --exclude ./lib/*
 	python3 -m unittest tests/*.py
 
 clean:
-	rm -rf ./sdx-common
\ No newline at end of file
+	rm -rf ./sdx-common
```
Comment (JamesGardiner) [inline Makefile]: Remove

File: app/main.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+#   encoding: UTF-8
+
+import logging
+import os.path
+import sys
+
+from sdc.rabbit import MessageConsumer
+from sdc.rabbit import QueuePublisher
+
+sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), "..")))
```
Comment (JamesGardiner) [inline app/main.py]: If we used explicit relative imports this wouldn't be needed

File: app/main.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+from app.response_processor import ResponseProcessor
+import app.settings
+
+
+def run():
+    logging.basicConfig(format=app.settings.LOGGING_FORMAT)
```
Comment (JamesGardiner) [inline app/main.py]: Same here - do `from . import settings` and access settings directly rather than doing path manipulation

File: app/response_processor.py [hunk 10]
```diff
@@ -1,16 +1,25 @@
 from json import dumps
+import logging
 import os
 
 from requests.packages.urllib3.exceptions import MaxRetryError
+from sdc.rabbit.exceptions import BadMessageError, PublishMessageError, RetryableError
```
Comment (JamesGardiner) [inline app/response_processor.py]: Some actions will probably need QuarantinableError

File: app/response_processor.py [hunk 11]
```diff
@@ -1,16 +1,25 @@
...
+from structlog import wrap_logger
 
 from app import settings
```
Comment (JamesGardiner) [inline app/response_processor.py]: Same here - use explicit relative over this

File: app/response_processor.py [hunk 12]
```diff
@@ -1,16 +1,25 @@
...
 from app.private_publisher import PrivatePublisher
 from app.settings import session
-from app.helpers.exceptions import DecryptError, BadMessageError, RetryableError
+
+logger = wrap_logger(logging.getLogger(__name__))
 
 
 class ResponseProcessor:
 
+    @classmethod
```
Comment (JamesGardiner) [inline app/response_processor.py]: This seems obtuse, considering we just need to supply a callable - what is the benefit over instantiating an instance of this class, and accessing the process method via it?

File: app/response_processor.py (hunk 3 above)
Comment (JamesGardiner) [inline app/response_processor.py]: `self.logger.error("No valid service name", exception=e)` would be better

File: app/response_processor.py (hunk 5 above)
Comment (JamesGardiner) [inline app/response_processor.py]: Talking about bound loggers, doesn't it need to be unbound once it has finished dealing with this message?

File: requirements.txt [hunk 13]
```diff
@@ -2,6 +2,7 @@ cryptography==1.7.1
 requests==2.12.4
 pika==0.10.0
 Jinja2==2.8
+sdc-rabbit==0.3.4
 sdx-common==0.7.0
```
Comment (JamesGardiner) [inline requirements.txt]: Is this defunct in this package now? I can't see where it is used

File: app/response_processor.py [hunk 14]
```diff
@@ -46,41 +52,44 @@ def service_name(self, url=None):
...
+            self.logger.error("No valid service name", exception=e)
 
-    def process(self, encrypted_survey):
-        # decrypt
+    def _process(self, encrypted_survey):
         decrypted_json = self.decrypt_survey(encrypted_survey)
 
         metadata = decrypted_json['metadata']
-        self.logger = self.logger.bind(user_id=metadata['user_id'], ru_ref=metadata['ru_ref'])
+        self.logger = self.logger.bind(user_id=metadata['user_id'],
```
Comment (JamesGardiner) [inline app/response_processor.py]: Use the `get` method on all the dictionary access calls here to avoid `KeyError` exceptions when a malformed message is passed.

File: app/response_processor.py [hunk 15]
```diff
@@ -46,41 +52,44 @@ def service_name(self, url=None):
...
+                                       ru_ref=metadata['ru_ref'])
 
         if not self.tx_id:
             self.tx_id = decrypted_json.get('tx_id')
         elif self.tx_id != decrypted_json.get('tx_id'):
+            logger.info('tx_ids from decrypted_json and message header do not match.' +
+                        ' Rejecting message',
+                        decrypted_tx_id=decrypted_json.get('tx_id'),
+                        message_tx_id=self.tx_id)
             raise BadMessageError
```
Comment (JamesGardiner) [inline app/response_processor.py]: We should probably quarantine these messages

File: app/response_processor.py [hunk 16]
```diff
@@ -46,105 +52,129 @@ def service_name(self, url=None):
...
-            # only send the receipt is the json is valid
-            if decrypted_json.get("survey_id") != "feedback":
-                self.logger.info("Receipting survey", tx_id=self.tx_id)
-                self.send_receipt(decrypted_json)
-            else:
-                self.logger.info("Feedback survey, skipping receipting", tx_id=self.tx_id)
-        finally:
-            # store the survey regardless
-            self.store_survey(decrypted_json)
+            self.logger.info("Feedback survey, skipping receipting")
```
Comment (warrenbailey) [inline app/response_processor.py]: If the survey is invalid then this will log out "Feedback survey, skipping receipting" even if it isn't the feedback survey. I think the logic was clearer before, please change back

File: app/response_processor.py [hunk 17]
```diff
@@ -46,105 +52,129 @@ def service_name(self, url=None):
...
 
-        return
+        self.store_survey(decrypted_json)
+
+        self.logger.unbind("user_id", "ru_ref", "tx_id")
 
     def send_receipt(self, decrypted_json):
         receipt_json = {
-            'tx_id': decrypted_json['tx_id'],
+            'tx_id': decrypted_json.get('tx_id'),
```
Comment (warrenbailey) [inline app/response_processor.py]: I think this might be better wrapped in a try/except KeyError block, because these values should be present in every case.

File: app/response_processor.py [hunk 18]
```diff
@@ -46,105 +52,129 @@ def service_name(self, url=None):
...
     def store_survey(self, decrypted_json):
         self.logger.info("Storing survey")
-        self.response_ok(self.remote_call(settings.SDX_RESPONSES_URL, json=decrypted_json))
+        response = self.remote_call(settings.SDX_RESPONSES_URL,
+                                    json=decrypted_json)
+        try:
+            self.response_ok(response)
+        except ClientError:
+            self.logger.error("Survey storage unsuccessful. Quarantining Survey.")
+            raise QuarantinableError
```
Comment (warrenbailey) [inline app/response_processor.py]: This is correct? We're quarantining the responses if we can't store it?

In that case I think we should be retrying, what was the logic before?

File: app/settings.py [hunk 19]
```diff
@@ -5,7 +5,8 @@
 from requests.adapters import HTTPAdapter
 
 
-LOGGING_LEVEL = logging.getLevelName(os.getenv('LOGGING_LEVEL', 'DEBUG'))
+LOGGING_LEVEL = logging.getLevelName(os.getenv("LOGGING_LEVEL", "DEBUG"))
+LOGGING_FORMAT = "%(asctime)s.%(msecs)06dZ|%(levelname)s: sdx-collect: %(message)s"
```
Comment (insacuri) [inline app/settings.py]: Does this include the T in between the yyyy-m-ddThh:MM:ss.mmmmmmZ? (with the M being minutes and m being milliseconds)

File: tests/test_response_processor.py [hunk 20]
```diff
@@ -2,10 +2,13 @@
 import json
 import logging
 import os
+import responses
```
Comment (JamesGardiner) [inline tests/test_response_processor.py]: Imports should be in 3 sections - stdlib, packages, local packages/modules

File: app/settings.py (hunk 19 above)
Comment (tej99) [inline app/settings.py]: The date format is set in main.py. 'datefmt="%Y-%m-%dT%H:%M:%S"' does not have milliseconds.

File: app/response_processor.py (hunk 18 above)
Comment (tej99) [inline app/response_processor.py]: It was throwing a BadMessageError if store gives back 400 response. I don't think it should be retrying once sdx-store gives a 400 response and neither should it be throwing it away because it has gone through decrypt and validate.

File: app/response_processor.py (hunk 17 above)
Comment (tej99) [inline app/response_processor.py]: Should I just quarantine it if tx_id is missing at this point or should it carry on processing and got to store?

File: app/response_processor.py [hunk 21]
```diff
@@ -21,19 +31,15 @@ def options():
             pass
```
Comment (tej99) [inline app/response_processor.py]: no need to catch KeyError*
